### ISSUE TITLE
Oso should use sqlalchemy.orm.aliased on tables involved in permissions

### ISSUE THREAD

**Issue by lafrech** [osohq/oso#1633]
Kind of a follow-up to #1531 but I figured it deserved a separate issue. Let's keep #1531 for the `table name "xxx" specified more than once` issue.

As written there, when joining tables to a oso authorized query, I need to alias the tables I join to in case they are used internally by oso to check the permissions. I only need to do it if the table is actually used, but when dealing with a specific part of the code, I'm not supposed to know that, and it might change in the future, so I'm aliasing every table. Not very convenient.

This could be avoided if oso just aliased all the tables it uses. I don't think there's any drawback to that, is there?

Also, I just got hit by another consequence of this. While trying to use joined table inheritance, I got [this warning](https://docs.sqlalchemy.org/en/14/errors.html#an-alias-is-being-generated-automatically-due-to-overlapping-tables):

> An alias is being generated automatically due to overlapping tables

The docs explain I should be aliasing the tables I join to queries on joined tables to avoid this. The whole thing is still a bit blurry to me TBH but I'm pretty sure that aliasing all tables in oso generated joins would fix this.

I wanted to try this but couldn't as apparently it is not as simple as `pip install -e`, there are steps involved to get header files in the right place.

```py
class SqlAlchemyAdapter(DataAdapter):
    def __init__(self, session: Session) -> None:
        self.session = session

    def build_query(self, filter):
        types = filter.types

        def re(q, rel):
            typ = types[rel.left]
            rec = typ.fields[rel.name]

            # ------------------------------------------
            # Alias tables involved in the query
            # left = typ.cls
            # right = types[rec.other_type].cls
            left = aliased(typ.cls)  # Or should it be aliased(typ.cls, flat=True) ?
            right = aliased(types[rec.other_type].cls)
            # ------------------------------------------

            return q.join(
                right, getattr(left, rec.my_field) == getattr(right, rec.other_field)
            )
```

**Comment by lafrech** [comment]
The naive approach above is wrong.

I took a bit of time to experiment with this and I'm beginning to understand why "just aliasing" is unfortunately not that simple. Giving up on this for now.

**Comment by lafrech** [comment]
Well, without taking time to further work on this, I'm afraid all I'll be able to write here will be close to a big pile of nonsense.

This said, here's what I understand:

The SQLA adapter takes each relation involved in determining the permissions one after another and adds the corresponding join to the base query. Basically.

Aliasing tables at each iteration (each relation) is wrong because since joins are chained, left table in relation N must be the same as right table in relation N+1, not another alias. Again, this is the rough idea.

To get things to work, you'd need to ensure that table is aliased to the same alias. So you can't just treat relations one after the other. You need to work one level above the `re` function quoted in my comment above, in the `build_query` function, to deal with this. And even then, I don't know if we have enough information at this stage to decide, for a given table appearing multiple times, which occurrences belong to the same alias.

Basically, if the only key here is the table itself, there is no way to tell. This would explain the current limitation of one appearance per table in a relation.

This is my totally naive take on this. Developers may want to correct this and elaborate.

I can't tell if this issue should remain open. I'd say it might help users trying to investigate this (but this is true even if it is closed). The point of opening this was that I thought the aliasing thing could be much simpler to address, but I'm beginning to think it all boils down to the limitation explained in #1531.

**Comment by killpack** [comment]
I appreciate all that detail! Since it seems like this is proposing a solution to #1531 rather than raising a new issue (and, from your investigation, the proposed solution didn't really work out), I'm going to close this issue. Thanks again for taking the time to look into this and to write down what you learned.
